### PR TITLE
Laboratory work #1, Eva Patrakova - 19FPL2

### PR DESCRIPTION
**Pull request by EvaPatrakova**

### CODE REVIEW

**Comment by ValeriyaKuznetsova** [comment]
@EvaPatrakova 
Please, rename your Pull Request. Use this template – Laboratory work #1, Fedorov Makar - 19FPL1

**Comment by ValeriyaKuznetsova** [comment]
Unfortunately, the deadline is not that far. Please, push changes so that we can review them.

File: lab_1/main.py [hunk 1]
```diff
@@ -3,17 +3,33 @@
...
-    """
-    Splits sentences into tokens, converts the tokens into lowercase, removes punctuation
-    :param text: the initial text
-    :return: a list of lowercased tokens without punctuation
-    e.g. text = 'The weather is sunny, the man is happy.'
-    --> ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
-    """
-    pass
-
+text = open('data.txt', encoding='utf-8').read()
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: remove lines 6,7, 31, 32
1) you should do this in start.py
2) a function cannot be into if/else, it should exist on its own

**the logic of main.py**
def ...():
    ...

def ...():
    ...

def ...():
    ...

So, there should be only functions and their implementations

File: lab_1/main.py [hunk 2]
```diff
@@ -3,17 +3,33 @@
...
+            text = text.replace('\n', ' ')
+        if '\t' in text:
+            text = text.replace('\t', ' ')
+        for i in text:
+            if i.isalpha() or i == ' ':
+                tokens += i
+        tokens = tokens.split(' ')
+        while '' in tokens:
+            tokens.remove('')
+        print(tokens)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: remove prints (everywhere)

File: lab_1/main.py [hunk 3]
```diff
@@ -41,21 +38,41 @@ def remove_stop_words(tokens: list, stop_words: list) -> list:
...
     """
-    pass
+    without_stop_words = []
+    stop_words2 = ''
+    if isinstance(tokens, list) and len(tokens) != 0 and len(stop_words) != 0 and isinstance(stop_words, list):
+        for n in tokens:
+            if n not in stop_words:
+                without_stop_words.append(n)
+        return without_stop_words
+    elif type(tokens) != list or len(tokens) == 0:
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: move lines 48 and 50 to the beginning

File: lab_1/main.py [hunk 4]
```diff
@@ -65,10 +82,19 @@ def get_top_n_words(freq_dict: dict, top_n: int) -> list:
     top_n = 1
     --> ['happy']
     """
-    pass
+    top_n_words = []
+    if isinstance(freq_dict, dict) and isinstance(top_n, int):
+        freq_dict = list(freq_dict.items())
+        freq_dict.sort(reverse=True)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: try this: freq_dict = {'a': 100, 'z': 2}
and this: freq_dict = {'a': 2, 'z': 100}

the result should be different

File: lab_1/main.py [hunk 5]
```diff
@@ -3,16 +3,15 @@
...
-if type(text) is str:
-    def tokenize(text):
-        """
-        Splits sentences into tokens, converts the tokens into lowercase, removes punctuation
-        :param text: the initial text
-        :return: a list of lowercased tokens without punctuation
-        e.g. text = 'The weather is sunny, the man is happy.'
-        --> ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
-        """
+def tokenize(text):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: return type annotation (eveywhere)
(text: str) -> list

**Review by ValeriyaKuznetsova — COMMENTED** [review]
correct all the issues and work on tests

**Comment by ValeriyaKuznetsova** [comment]
do not forget about target_score.txt and start.py

**Comment by EvaPatrakova** [comment]
I change some functions, but I have some problems with get_concordance function: it does not work with bad_input in the place of word and I do not have an idea how to fix it

**Comment by ValeriyaKuznetsova** [comment]
tests fail when left_context_size/right_context_size = True or False
just add simple check on this type

**Comment by EvaPatrakova** [comment]
For better checking some of the functions, I needed to fix some tests because there was wrong file path. Do I need to push new tests` programmes for it?

**Comment by ValeriyaKuznetsova** [comment]
do not push (if you mean "lab_1/")

**Comment by ValeriyaKuznetsova** [comment]
delete everything from target score and put 8 there

**Comment by ValeriyaKuznetsova** [comment]
Work on lint, make a pipeline green

File: lab_1/main.py [hunk 6]
```diff
@@ -87,16 +177,75 @@ def get_adjacent_words(tokens: list, word: str, left_n: int, right_n: int) -> li
...
+        all_adjacent_words.append(context)
+        if num > 1:
+            while num != 1:
+                concordance.remove(text)
+                new_text = concordance[0]
+                new_context = []
+                new_context.append(new_text[0])
+                all_adjacent_words.append(new_context)
+                num -= 1
+
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: remove this line, maybe this will help with lint

File: lab_1/main.py [hunk 7]
```diff
@@ -87,16 +177,75 @@ def get_adjacent_words(tokens: list, word: str, left_n: int, right_n: int) -> li
...
+    test4 = isinstance(right_n, int)
+    if not (test1 and test2 and word in tokens and test3 and test4):
+        return []
+    if isinstance(left_n, bool) or isinstance(right_n, bool):
+        return []
+    if right_n > 0 and left_n > 0:
+        concordance = get_concordance(tokens, word, left_n, right_n)
+        num = len(concordance)
+        text = concordance[0]
+        context.append(text[0])
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: if the later doesn't work, change this to extend with [text[0], text[-1]]
and remove 195

File: lab_1/main.py [hunk 8]
```diff
@@ -87,16 +177,75 @@ def get_adjacent_words(tokens: list, word: str, left_n: int, right_n: int) -> li
...
+        context.append(text[-1])
+        all_adjacent_words.append(context)
+        if num > 1:
+            concordance.remove(text)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: you do not really need all of this, you can iterate over a concordance itself using for

File: lab_1/main.py [hunk 9]
```diff
@@ -87,16 +177,75 @@ def get_adjacent_words(tokens: list, word: str, left_n: int, right_n: int) -> li
...
+            while num >= 1:
+                new_text = concordance[0]
+                new_context = []
+                new_context.append(new_text[0])
+                new_context.append(new_text[-1])
+                all_adjacent_words.append(new_context)
+                concordance.remove(new_text)
+                num = len(concordance)
+    elif right_n > 0:
+        concordance = get_concordance(tokens, word, left_n, right_n)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: and it's not good to have the same parts of code in one function

File: lab_1/main.py [hunk 10]
```diff
@@ -105,7 +254,13 @@ def write_to_file(path_to_file: str, content: list):
     """
     Writes the result in a file
     """
-    pass
+
+    with open(path_to_file, 'w') as report:
+        for text in content:
+            context = ' '.join(text) + '\n'
+            report.write(context)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: first, put each context in one variable
only then write it to file
it's not good to write to file at each step

File: lab_1/start.py [hunk 11]
```diff
@@ -2,17 +2,43 @@
...
+    print('Adjacent words for word "happy": {} '.format(adjacent_words))
 
-    #  here goes your logic: calling methods from concordance.py
+    write_to_file = write_to_file('report.txt', concordance)
 
-    RESULT = None
+    RESULT = write_to_file
     # DO NOT REMOVE NEXT LINE - KEEP IT INTENTIONALLY LAST
-    assert RESULT == [(), ()], 'Concordance not working'
+    assert RESULT, 'Concordance not working'
```
Comment (ValeriyaKuznetsova) [inline lab_1/start.py]: assert RESULT == [expected result]

for example, [['the', 'man']]

File: lab_1/main.py [hunk 12]
```diff
@@ -189,56 +185,22 @@ def get_adjacent_words(tokens: list, word: str, left_n: int, right_n: int) -> li
...
-            concordance.remove(text)
-            while num >= 1:
-                new_text = concordance[0]
-                new_context = []
-                new_context.append(new_text[0])
-                new_context.append(new_text[-1])
-                all_adjacent_words.append(new_context)
-                concordance.remove(new_text)
-                num = len(concordance)
+        print(concordance)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py]: remove print

**Comment by ValeriyaKuznetsova** [comment]
correct two issues (about RESULT and print)

File: lab_1/start.py (hunk 11 above)
Comment (EvaPatrakova) [reply lab_1/start.py]: Thank you very much for assisting!

File: lab_1/start.py [hunk 13]
```diff
@@ -2,17 +2,43 @@
...
+
+    adjacent_words = get_adjacent_words(['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy',
+                                         'the', 'dog', 'is', 'happy', 'but', 'the', 'cat', 'is', 'sad'], 'happy', 2, 3)
+    print('Adjacent words for word "happy": {} '.format(adjacent_words))
 
-    #  here goes your logic: calling methods from concordance.py
+    write_to_file = write_to_file('report.txt', concordance)
 
-    RESULT = None
+    RESULT = write_to_file
```
Comment (ValeriyaKuznetsova) [inline lab_1/start.py]: I think you meant adjacent_words

File: lab_1/start.py (hunk 11 above)
Comment (EvaPatrakova) [reply lab_1/start.py]: It needs to be printed with , 'Concordance not working ?

**Comment by ValeriyaKuznetsova** [comment]
Open another PR with the second lab

**Pull request closed without merging**